### PR TITLE
Support Segwit (including P2SH wrapped) signatures created by Electrum

### PR DESCRIPTION
**Pull request by lf94**
Should resolve the following issues:
* https://github.com/bitcoinjs/bitcoinjs-message/issues/28
* https://github.com/bitcoinjs/bitcoinjs-message/issues/20
* https://github.com/coreyphillips/moonshine/issues/24

Big thanks to @coreyphillips who I saw found the root cause. While I knew it was some offset happening, I was not sure why.

Sponsored by Bity and Relai.

### CODE REVIEW

**Comment by junderw** [comment]
Judging from #6 and #12 it seems that Electrum treats P2SH-P2WPKH as the exact same offset as P2PKH-COMPRESSED.

Since it is not a standard BIP, choosing one will break the other, unless Trezor decided to switch to a new format?

Since this Segwit support was modeled after Trezor, I'm going to ask for a review from 

Trezor:
@prusnak 

and

Electrum:
@ecdsa 
@SomberNight 
@lukechilds

**Comment by junderw** [comment]
Pasting the original issue from bitcoinjs-lib by @prusnak 

> For TREZOR we implemented a way how to Sign and Verify SegWit P2SH addresses. This is achieved by introducing extra `v` values used during signing.
> 
> Currently the following are used:
> 
> * 27-30 uncompressed pubkey (base58)
> * 31-34 compressed pubkey (base58)
> 
> We added the following:
> 
> * 35-38 p2sh segwit pubkey (base58)
> * 39-42 segwit pubkey (bech32)
> 
> [Unit tests for reference](https://github.com/trezor/python-trezor/blob/master/tests/device_tests/test_msg_signmessage_segwit.py)

Reference https://github.com/bitcoinjs/bitcoinjs-lib/issues/880

**Comment by junderw** [comment]
So by doing -4 for p2sh you are turning `35-38` into `31-34` which means that it is the same as P2PKH compressed.

So I think what happened was:

1. Core didn't support segwit or wrapped segwit
2. Trezor tried getting them to support it (?) but Core being Core, was taking forever
3. Trezor created their own method without a BIP
4. Core added support but did not change the flag byte
5. Electrum followed suit

Does this sound right?

File: index.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ function encodeSignature (signature, recovery, compressed, segwitType) {
   if (segwitType !== undefined) {
     recovery += 8
     if (segwitType === SEGWIT_TYPES.P2WPKH) recovery += 4
+    if (segwitType === SEGWIT_TYPES.P2SH_P2WPKH) recovery -= 4
```
Comment (junderw) [inline index.js:31]: This fixes the problem, but is obviously not the correct solution.

decodeSignature is not changed, so with this "fix" a p2sh-p2wpkh signed by this commit will not be able to be verified by this commit.

**Review by junderw — CHANGES_REQUESTED** [review]

**Comment by lf94** [comment]
@junderw your interpretation of the situation is absolutely correct. It only makes sense to continue with this implementation until a standard BIP arrives. If bitcoinjs-message continues to perpetuate this variant, it only harms the ecosystem further.

I would suggest a major version change if following semver.

**Comment by junderw** [comment]
Can someone link the PR where Bitcoin Core added message signing support for p2wpkh and p2sh-p2wpkh?

BitcoinJS should be following Core in absence of any BIP.

File: index.js (hunk 1 above)
Comment (lf94) [reply index.js:31]: I'll add some tests which use both the sign and verify to ensure they are sound.

**Comment by lf94** [comment]
> Can someone link the PR where Bitcoin Core added message signing support for p2wpkh and p2sh-p2wpkh?
> 
> BitcoinJS should be following Core in absence of any BIP.

It appears Bitcoin Core still does not support signing messages with P2SH or P2WPKH addresses.

**Comment by junderw** [comment]
Ah ok, so this is two non-core implementations battling it out... I see...

hmm...

I personally like Trezor's method better...

*sigh*

Maybe I can add the two following in a backwards compatible way.

1. Allow verify to succeed for either method.
2. Add (yet another) 2 segwit options for `p2sh(p2wpkh)-electrum` `p2wpkh-electrum`

ugh...

I would much rather just pick one or the other. In which case I'd pick Trezor since they were first and they actually removed ambiguity. (In the electrum method, you can't tell the script type from the flag byte, so you just have to try all the address types.)

File: index.js (hunk 1 above)
Comment (junderw) [reply index.js:31]: I would just wait.

I want to hear some feedback from Electrum / Trezor first before deciding what to do.

**Comment by lf94** [comment]
I 100% agree the situation (and the fact it's still a situation) is disappointing to say the least.

My vote's on adding an Electrum method then. When there's a standard BIP, maybe do a new major release?

File: index.js (hunk 1 above)
Comment (lf94) [reply index.js:31]: Sounds good, I'll wait. :slightly_smiling_face:

**Comment by SomberNight** [comment]
I had collected references/links here at one point: https://github.com/spesmilo/electrum/issues/3861

**Comment by junderw** [comment]
@lf94 I made some changes.

Could you read the README and try out signing and verifying using the methods described?

**Comment by junderw** [comment]
I have verified this pull request works as per the new instructions in the README against the latest version of Electrum.

**Review by junderw — APPROVED** [review]

**Comment by prusnak** [comment]
>  It only makes sense to continue with this implementation until a standard BIP arrives.

Standard BIP for this will never come.

BIP folks descided to come up with a completely new different standard which signs the whole scripts, not just the addresses - https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki

**Comment by ademcan** [comment]
Thank you very much for the PR @lf94 it solves my issue (#28 )! And thank you all for the great discussion

**Comment by lf94** [comment]
@junderw looks good to me. Thank you for your time! I guess we will add support for verifying both Electrum and Trezor generated signatures as you've done here.

**Pull request merged**